File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/planning/goal_tracking_controller.py
import logging
from django.db import transaction
from django.shortcuts import render, redirect, get_object_or_404
from django.contrib import messages
from django.http import HttpRequest, HttpResponse
from ....Models.companies.branch import Branch
from ....Models.individuals.employee import Employee
from ....Models.planning.goal_tracking import GoalTracking
from ....Models.planning.goal_type import GoalType
from .._traits.controller import Controller
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
import inspect

logger = logging.getLogger(__name__)

class GoalTrackingController(Controller):

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            return cls.__index_impl(request)
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def __index_impl(cls, request: HttpRequest) -> HttpResponse:
        if not request.user.has_perm('manage_goal_tracking'):
            return default_permission_denial(request, err=None, ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}', logger=logger)
        user = request.user
        if user.type == 'Employee':
            emp = Employee.objects.filter(user_id=user.id).first()
            qs = GoalTracking.objects.filter(created_by=user.id, branch=emp.branch_id)
        else:
            qs = GoalTracking.objects.filter(created_by=user.id)
        goal_trackings = qs.select_related('goal_type', 'branches')
        return render(request, 'goaltracking/index.html', {'goal_trackings': goal_trackings})

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            return cls.__create_impl(request)
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def __create_impl(cls, request: HttpRequest) -> HttpResponse:
        if not request.user.has_perm('create_goal_tracking'):
            return default_permission_denial(request, err=None, ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}', logger=logger)
        branches = Branch.objects.filter(created_by=request.user.id).values_list('id', 'name')
        goal_types = GoalType.objects.filter(created_by=request.user.id).values_list('id', 'name')
        status = GoalTracking.status
        return render(request, 'goaltracking/create.html', {
            'branches': branches,
            'goal_types': goal_types,
            'status': status,
        })

    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            with transaction.atomic():
                return cls.__store_impl(request)
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def __store_impl(cls, request: HttpRequest) -> HttpResponse:
        if not request.user.has_perm('create_goal_tracking'):
            return default_permission_denial(request, err=None, ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}', logger=logger)
        data = request.POST
        required = ('branch', 'goal_type', 'start_date', 'end_date', 'subject')
        if any(not data.get(f) for f in required):
            messages.error(request, 'Missing required fields.')
            return redirect('/')
        gt = GoalTracking()
        cls._set_goal_tracking(gt, data, request.user.id)
        gt.save()
        messages.success(request, 'Goal tracking successfully created.')
        return redirect('/')

    @classmethod
    def show(cls, request: HttpRequest, pk: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            gt = get_object_or_404(GoalTracking, pk=pk)
            return render(request, 'goaltracking/show.html', {'goal_tracking': gt})
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def edit(cls, request: HttpRequest, pk: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('edit_goal_tracking'):
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            gt = get_object_or_404(GoalTracking, pk=pk)
            branches = Branch.objects.filter(created_by=request.user.id).values_list('id', 'name')
            goal_types = GoalType.objects.filter(created_by=request.user.id).values_list('id', 'name')
            status = GoalTracking.status
            ratings = gt.rating
            return render(request, 'goaltracking/edit.html', {
                'branches': branches,
                'goal_types': goal_types,
                'goal_tracking': gt,
                'ratings': ratings,
                'status': status,
            })
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def update(cls, request: HttpRequest, pk: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            with transaction.atomic():
                return cls.__update_impl(request, pk)
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def __update_impl(cls, request: HttpRequest, pk: int) -> HttpResponse:
        if not request.user.has_perm('edit_goal_tracking'):
            return default_permission_denial(request, err=None, ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}', logger=logger)
        gt = get_object_or_404(GoalTracking, pk=pk)
        data = request.POST
        required = ('branch', 'goal_type', 'start_date', 'end_date', 'subject')
        if any(not data.get(f) for f in required):
            messages.error(request, 'Missing required fields.')
            return redirect('/')
        cls._set_goal_tracking(gt, data, request.user.id)
        gt.status = data.get('status')
        gt.progress = data.get('progress')
        gt.rating = data.get('rating')
        gt.save()
        messages.success(request, 'Goal tracking successfully updated.')
        return redirect('/')

    @classmethod
    def destroy(cls, request: HttpRequest, pk: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            with transaction.atomic():
                return cls.__destroy_impl(request, pk)
        except Exception as ex:
            return default_undefined_exception(request, err=ex, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def __destroy_impl(cls, request: HttpRequest, pk: int) -> HttpResponse:
        if not request.user.has_perm('delete_goal_tracking'):
            return default_permission_denial(request, err=None, ref=f'{cls.__name__}::{inspect.currentframe().f_code.co_name}', logger=logger)
        gt = get_object_or_404(GoalTracking, pk=pk)
        if gt.created_by == request.user.id:
            gt.delete()
            messages.success(request, 'Goal tracking successfully deleted.')
        else:
            messages.error(request, 'Permission denied.')
        return redirect('/')

    @classmethod
    def _set_goal_tracking(cls, obj: GoalTracking, data, user_id: int):
        obj.branch = data.get('branch')
        obj.goal_type = data.get('goal_type')
        obj.start_date = data.get('start_date')
        obj.end_date = data.get('end_date')
        obj.subject = data.get('subject')
        obj.target_achievement = data.get('target_achievement')
        obj.description = data.get('description')
        obj.created_by = user_id

File: _inc/laravel/app/Http/Controllers/Planning/GoalTrackingController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    CompaniesConstants,
    DatabaseConstants,
    PermissionsConstants,
    UsersConstants,
    ViewsConstants
};
use App\Models\{
    Branch,
    Employee,
    GoalTracking,
    GoalType
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\{
    Log,
    Validator,
    View as ViewFacade
};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
final class GoalTrackingController extends Controller
{
    use HasCrudConstants;

    use ChecksLogin, ChecksPermissions;
    private const REDIRECT_INDEX = '/';

    public function index(Request $request): RedirectResponse|View
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::GL_TRC . '.index';

        return $this->measureProfile($action, function () use ($request, $view, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, PermissionsConstants::MNG_GTR, self::REDIRECT_INDEX)) !== true) return $resp;
                $user = $userOrRedirect;
                $goalTrackings = strtolower($user[UsersConstants::COL_TP]) === 'employee'
                    ? GoalTracking::with(['goalType', 'branch'])
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->where('branch', Employee::where(UsersConstants::COL_USER_ID, $user?->id)->value(CompaniesConstants::COL_BRC_ID))
                    ->get()
                    : GoalTracking::with(['goalType', 'branch'])
                    ->where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->get();
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \RuntimeException('View not found'), $action);
                return view($view, compact('goalTrackings'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    public function create(Request $request): RedirectResponse|View
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::GL_TRC . '.create';

        return $this->measureProfile($action, function () use ($request, $view, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, 'create goal tracking', self::REDIRECT_INDEX)) !== true) return $resp;
                $user = $userOrRedirect;
                $branches = Branch::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck(UsersConstants::COL_NM, 'id')->prepend('Select Branch', '');
                $goalTypes = GoalType::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck(UsersConstants::COL_NM, 'id')->prepend('Select Goal Type', '');
                $status = GoalTracking::$status;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \RuntimeException('View not found'), $action);
                return view($view, compact('branches', 'goalTypes', 'status'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    public function store(Request $request): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, 'create goal tracking', self::REDIRECT_INDEX)) !== true) return $resp;
                if ($err = self::validateInput($request, [
                    'branch'     => 'required',
                    'goal_type'  => 'required',
                    'start_date' => 'required|date',
                    'end_date'   => 'required|date|after_or_equal:start_date',
                    'subject'    => 'required',
                ])) return $err;
                $user = $userOrRedirect;

                GoalTracking::create([
                    'branch'             => $request->branch,
                    'goal_type'          => $request->goal_type,
                    'start_date'         => $request->start_date,
                    'end_date'           => $request->end_date,
                    'subject'            => $request->subject,
                    'target_achievement' => $request->target_achievement,
                    'description'        => $request->description,
                    DatabaseConstants::COL_TABLE_CREATOR => $user?->creatorId(),
                ]);

                return redirect()->route(ViewsConstants::GL_TRC . '.index')
                    ->with('success', __('Goal tracking successfully created.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    public function show(Request $request, GoalTracking $goalTracking): RedirectResponse|View|null
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::GL_TRC . '.show';

        return $this->measureProfile($action, function () use ($request, $goalTracking, $view, $action) {
            Log::info("$action started", [UsersConstants::COL_USER_ID => $request->user()->id, 'goal_tracking_id' => $goalTracking->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($resp = self::guard($request, 'view goal tracking', self::REDIRECT_INDEX)) !== true) {
                Log::warning("$action permission denied", [UsersConstants::COL_USER_ID => $user?->id]);
                return $resp;
            }
            if ($goalTracking[DatabaseConstants::COL_TABLE_CREATOR] !== $user?->creatorId()) {
                Log::warning("$action ownership denied", [UsersConstants::COL_USER_ID => $user?->id, 'goal_tracking_id' => $goalTracking->id]);
                return defaultPermissionDenial($request, new AuthorizationException(), $action, route(self::REDIRECT_INDEX));
            }
            try {
                Log::info("$action rendering view", ['goal_tracking_id' => $goalTracking->id]);
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \RuntimeException('View not found'), $action, route(self::REDIRECT_INDEX));
                return view($view, compact('goalTracking'));
            } catch (\Throwable $e) {
                Log::error("$action failed", ['error' => $e->getMessage(), 'goal_tracking_id' => $goalTracking->id]);
                return defaultUndefinedException($request, $e, $action, route(self::REDIRECT_INDEX));
            }
        });
    }

    public function edit(Request $request, int|string $id): RedirectResponse|View
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::GL_TRC . '.edit';

        return $this->measureProfile($action, function () use ($request, $id, $view, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, 'edit goal tracking', self::REDIRECT_INDEX)) !== true) return $resp;
                $user = $userOrRedirect;
                $goalTracking = GoalTracking::findOrFail($id);
                if ($goalTracking[DatabaseConstants::COL_TABLE_CREATOR] !== $user?->creatorId())
                    return defaultPermissionDenial($request, new AuthorizationException(), $action);
                $branches = Branch::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck(CompaniesConstants::COL_BRC_NM, 'id')->prepend('Select Branch', '');
                $goalTypes = GoalType::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->pluck('name', 'id')->prepend('Select Goal Type', '');
                $status = GoalTracking::$status;
                $ratings = json_decode($goalTracking->rating, true);
                if (!ViewFacade::exists($view)) return defaultUndefinedException($request, new \RuntimeException('View not found'), $action);
                return view($view, compact('branches', 'goalTypes', 'goalTracking', 'ratings', 'status'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    public function update(Request $request, int|string $id): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request, $id, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, 'edit goal tracking', self::REDIRECT_INDEX)) !== true) return $resp;
                if ($err = self::validateInput($request, [
                    'branch'     => 'required',
                    'goal_type'  => 'required',
                    'start_date' => 'required|date',
                    'end_date'   => 'required|date|after_or_equal:start_date',
                    'subject'    => 'required',
                ])) return $err;
                $user = $userOrRedirect;

                $gt = GoalTracking::findOrFail($id);
                if ($gt[DatabaseConstants::COL_TABLE_CREATOR] !== $user?->creatorId())
                    return defaultPermissionDenial($request, new AuthorizationException(), $action);

                $gt->fill([
                    'branch'             => $request->branch,
                    'goal_type'          => $request->goal_type,
                    'start_date'         => $request->start_date,
                    'end_date'           => $request->end_date,
                    'subject'            => $request->subject,
                    'target_achievement' => $request->target_achievement,
                    'status'             => $request->status,
                    'progress'           => $request->progress,
                    'description'        => $request->description,
                    'rating'             => json_encode($request->rating),
                ])->save();

                return redirect()->route(ViewsConstants::GL_TRC . '.index')
                    ->with('success', __('Goal tracking successfully updated.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    public function destroy(Request $request, int|string $id): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        return $this->measureProfile($action, function () use ($request, $id, $action) {
            try {
                if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
                if (($resp = self::guard($request, 'delete goal tracking', self::REDIRECT_INDEX)) !== true) return $resp;
                $user = $userOrRedirect;

                $gt = GoalTracking::findOrFail($id);
                if ($gt[DatabaseConstants::COL_TABLE_CREATOR] !== $user?->creatorId())
                    return defaultPermissionDenial($request, new AuthorizationException(), $action);

                $gt->delete();
                return redirect()->route(ViewsConstants::GL_TRC . '.index')
                    ->with('success', __('Goal tracking successfully deleted.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($request, $e, $action);
            }
        });
    }

    private static function validateInput(Request $request, array $rules): ?RedirectResponse
    {
        $v = Validator::make($request->all(), $rules);
        return $v->fails()
            ? redirect()->back()->with('error', $v->errors()->first())
            : null;
    }
}
